src/driver/src/main/java/com/edgedb/driver/binary/PacketSerializer.java

```diff
@@ -187,24 +194,125 @@ protected void encode(@NotNull ChannelHandlerContext ctx, @NotNull Sendable msg,
 
     public static @Nullable Receivable deserialize(ServerMessageType messageType, long length, @NotNull ByteBuf buffer) {
         var reader = new PacketReader(buffer);
+        return deserializeSingle(messageType, length, reader, true);
+    }
+
+    public static @Nullable Receivable deserializeSingle(PacketReader reader) {
+        var messageType = reader.readEnum(ServerMessageType.class, Byte.TYPE);
+        var length = reader.readUInt32().longValue();
+
+        return deserializeSingle(messageType, length, reader, false);
+    }
 
-        if(!deserializerMap.containsKey(messageType)) {
-            logger.error("Unknown packet type {}", messageType);
-            reader.skip((int)length);
+    public static @Nullable Receivable deserializeSingle(
+            ServerMessageType type, long length, @NotNull PacketReader reader,
+            boolean verifyEmpty
+    ) {
+        if(!deserializerMap.containsKey(type)) {
+            logger.error("Unknown packet type {}", type);
+            reader.skip(length);
             return null;
         }
 
         try {
-            return deserializerMap.get(messageType).apply(reader);
+            return deserializerMap.get(type).apply(reader);
         }
         catch (Exception x) {
             logger.error("Failed to deserialize packet", x);
             throw x;
         }
         finally {
             // ensure we read the entire packet
-            if(!reader.isEmpty()) {
-                logger.warn("Hanging data left inside packet reader of type {} with length {}", messageType, length);
+            if(verifyEmpty && !reader.isEmpty()) {
+                logger.warn("Hanging data left inside packet reader of type {} with length {}", type, length);
+            }
+        }
+    }
+
+    public static HttpResponse.BodyHandler<List<Receivable>> PACKET_BODY_HANDLER = new PacketBodyHandler();
+
+    private static class PacketBodyHandler implements HttpResponse.BodyHandler<List<Receivable>> {
+        @Override
+        public HttpResponse.BodySubscriber<List<Receivable>> apply(HttpResponse.ResponseInfo responseInfo) {
+            // ensure success
+            var isSuccess = responseInfo.statusCode() / 100 == 2;
```

Comment:
Not sure how this works in the HTTP client we're using here, but would a 3XX response also be considered a success? Or does redirection happen implicitly by the client and you end up with 2XX/4XX/5XX?

Reply:
I believe redirection happens implicitly, although should we even support/handle redirects though in bindings @1st1?